app/jobs/application_job.rb

```diff
@@ -21,16 +20,69 @@ class ApplicationJob < ActiveJob::Base
     throw(:abort) unless job.class.before_enqueue_lock(*job.arguments)
   end
 
-  # Overriding so that changes in ActiveModel object don't result in new lock (which they do when just calling to_s).
-  def self.lock(*args)
-    new_args = args.map { |arg| arg.class.method_defined?(:to_global_id) ? arg.to_global_id.to_s : arg }
-    "lock:#{self.class.name}-#{new_args}"
+  around_perform do |job, block|
+    block.call
+  ensure
+    # Always clear the lock when we're done, even if there was an error.
+    job.class.clear_lock(*job.arguments)
   end
 
   # Override in subclass to tune per queue.
-  # def lock_timeout
-  #   3600
-  # end
+  # @return [Integer] lock_timeout in seconds, after which lock is no longer valid,
+  # regardless of whether the job that created the lock is still in queue or working.
+  def self.lock_timeout
+    # 1 hour is a reasonable default, because most jobs in this app will
+    # cause little if re-executed too aggressively.
+    3600
+  end
+
+  # @return [String] the key for locking this job/payload combination, e.g. 'lock:MySpecificJob-bt821jk7040;1'
+  def self.queue_lock_key(*args)
+    # Changes in ActiveModel object don't result in new lock (which they do when just calling to_s).
+    queue_lock_args = args.map { |arg| arg.class.method_defined?(:to_global_id) ? arg.to_global_id.to_s : arg }
+    "lock:#{name}-#{queue_lock_args.join(';')}"
+  end
+
+  def self.before_enqueue_lock(*args) # rubocop:disable Metrics/AbcSize
+    key = queue_lock_key(*args)
+    now = Time.now.to_i
+    new_expiry_time = now + lock_timeout + 1
+
+    # return true if we successfully acquired the lock
+    # "Set key to hold string value if key does not exist" (otherwise no-op) -- https://redis.io/commands/setnx
+    if Resque.redis.setnx(key, new_expiry_time)
+      Rails.logger.info("acquired lock on #{key} (none existed)")
+      return true
+    end
+
+    # see if the existing lock is still valid and return false if it is
+    # (we cannot acquire the lock during the timeout period)
+    key_expires_at = Resque.redis.get(key).to_i
+    if now <= key_expires_at
+      Rails.logger.info("failed to acquire lock on #{key}, because it has not expired (#{now} <= #{key_expires_at})")
+      return false
+    end
+
+    # otherwise set the new_expiry_time and ensure that no other worker has
+    # acquired the lock, possibly pushing out the expiry time further
+    # "Atomically sets key to value and returns the old value stored at key." -- https://redis.io/commands/getset
+    key_expires_at = Resque.redis.getset(key, new_expiry_time).to_i
+    if now > key_expires_at
+      Rails.logger.info("acquired lock on #{key} (old lock expired, #{now} > #{key_expires_at})")
+      true
+    else
+      Rails.logger.info("failed to acquired lock on #{key} but updated expiry time to #{new_expiry_time} (#{now} <= #{key_expires_at})")
+      false
```

Comment:
this `else` is the only part of the PR that's uncovered by tests, and that coverage deficit is what's failing the build.

maybe i'm missing something, but i think this is a hard case to test for, because it requires that a second worker pick up the lock in between the `get` from Redis on line 60 and the `getset` to Redis on line 69.  i do think this condition is worth covering, though, and i've handled it in the same way as the original resque-lock gem, albeit with a bit more logging detail.  but the logic should be equivalent (or that was the intent).

Reply:
i'm fine with this - this isn't the only else clause untested in the portfolio.